src/main/java/com/artipie/asto/bench/BenchmarkStorage.java

```diff
@@ -0,0 +1,144 @@
+    /**
+     * Backend storage.
+     */
+    private final InMemoryStorage backend;
+
+    /**
+     * Local storage.
+     */
+    private final Map<Key, byte[]> local;
+
+    /**
+     * Set which contains deleted keys.
+     */
+    private final Set<Key> deleted;
+
+    /**
+     * Ctor.
+     * @param backend Backend storage
+     */
+    public BenchmarkStorage(final InMemoryStorage backend) {
+        this(backend, Collections.emptyMap());
+    }
+
+    /**
+     * Ctor.
+     * @param backend Backend storage
+     * @param local Content of local storage
+     */
+    public BenchmarkStorage(final InMemoryStorage backend, final Map<Key, byte[]> local) {
+        this.backend = backend;
+        this.local = new ConcurrentSkipListMap<>(local);
+        this.deleted = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        final CompletableFuture<Void> res;
+        if (Key.ROOT.equals(key)) {
+            res = new CompletableFutureSupport.Failed<Void>(
+                new ArtipieIOException("Unable to save to root")
+            ).get();
+        } else {
+            res = new Concatenation(new OneTimePublisher<>(content)).single()
+                .to(SingleInterop.get())
+                .thenApply(Remaining::new)
+                .thenApply(Remaining::bytes)
+                .thenAccept(bytes -> this.local.put(key, bytes))
+                .toCompletableFuture();
+        }
+        return res;
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Long> size(final Key key) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Content> value(final Key key) {
+        return CompletableFuture.runAsync(
+            () -> {
+                if (this.deleted.contains(key)) {
+                    throw new ValueNotFoundException(key);
+                }
+            }
+        ).thenCompose(
+            noth -> {
+                final byte[] lcl = this.local.getOrDefault(key, null);
+                final CompletableFuture<Content> res;
+                if (lcl == null) {
+                    res = this.backend.value(key);
+                } else {
+                    res = CompletableFuture.completedFuture(
+                        new Content.OneTime(new Content.From(lcl))
+                    );
+                }
+                return res;
+            }
```

Comment:
And same here: we can access byte-array reference from map by key almost immediately, the async operation may start longer than map access time. And the back-end storage memory could be accessed directly via `Map<Key, byte[]>` passed to the back-end constructor (without async call)